Cleanup BodyCache

Nashenas88 opened the pull request:
After this PR:

- `BodyCache` is renamed to `BodyAndCache`
- `ReadOnlyBodyCache` is renamed to `ReadOnlyBodyAndCache`
- `ReadOnlyBodyAndCache::body` fn is removed and all calls to it are replaced by a deref (possible due to fix of its `Deref` imp in #65947)

cc @eddyb @oli-obk

rust-highfive commented:
r? @davidtwco

(rust_highfive has picked a reviewer for you, use r? to override)

Nashenas88 commented:
r? @eddyb

Nashenas88 commented:
This should address @eddyb 's comments here https://github.com/rust-lang/rust/pull/64736#discussion_r353097802 (and the comments below in that link).

Nashenas88 commented on src/librustc/mir/cache.rs:
```diff
@@ -220,10 +220,6 @@ impl ReadOnlyBodyCache<'a, 'tcx> {
         self.cache.unwrap_predecessor_locations(loc, self.body)
     }
 
-    pub fn body(&self) -> &'a Body<'tcx> {
```
@eddyb Thank you so much for pointing on the `&'a` in the `Deref` impl. That was the only reason this fn was left around, and it helped clean up a lot of code. CC @oli-obk

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=15106&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/15106/logs/65), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/15106/logs/65)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-12-03T16:57:39.4219675Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-12-03T16:57:40.3051365Z ##[command]git config gc.auto 0
2019-12-03T16:57:40.3058413Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-12-03T16:57:40.3064201Z ##[command]git config --get-all http.proxy
2019-12-03T16:57:40.3069056Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/66991/merge:refs/remotes/pull/66991/merge
---
2019-12-03T17:03:35.2079995Z    Compiling serde_json v1.0.40
2019-12-03T17:03:36.8525170Z    Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
2019-12-03T17:03:47.5697799Z     Finished release [optimized] target(s) in 1m 25s
2019-12-03T17:03:47.5796719Z tidy check
2019-12-03T17:03:48.2323409Z tidy error: /checkout/src/librustc_mir/util/def_use.rs:3: line longer than 100 chars
2019-12-03T17:03:48.2767910Z tidy error: /checkout/src/librustc_mir/shim.rs:322: line longer than 100 chars
2019-12-03T17:03:50.2831618Z some tidy checks failed
2019-12-03T17:03:50.2833957Z Found 486 error codes
2019-12-03T17:03:50.2834652Z Found 0 error codes with no tests
2019-12-03T17:03:50.2834751Z Done!
2019-12-03T17:03:50.2834751Z Done!
2019-12-03T17:03:50.2834854Z 
2019-12-03T17:03:50.2839001Z 
2019-12-03T17:03:50.2840118Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor"
2019-12-03T17:03:50.2840554Z 
2019-12-03T17:03:50.2840685Z 
2019-12-03T17:03:50.2849872Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
2019-12-03T17:03:50.2850203Z Build completed unsuccessfully in 0:01:29
2019-12-03T17:03:50.2850203Z Build completed unsuccessfully in 0:01:29
2019-12-03T17:03:50.2906645Z == clock drift check ==
2019-12-03T17:03:50.2918001Z   local time: Tue Dec  3 17:03:50 UTC 2019
2019-12-03T17:03:50.4479048Z   network time: Tue, 03 Dec 2019 17:03:50 GMT
2019-12-03T17:03:50.4482466Z == end clock drift check ==
2019-12-03T17:03:51.7940227Z 
2019-12-03T17:03:51.8012439Z ##[error]Bash exited with code '1'.
2019-12-03T17:03:51.8041518Z ##[section]Starting: Checkout
2019-12-03T17:03:51.8043904Z ==============================================================================
2019-12-03T17:03:51.8043987Z Task         : Get sources
2019-12-03T17:03:51.8044041Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

eddyb commented on src/librustc/mir/cache.rs:
```diff
@@ -272,25 +268,17 @@ impl graph::WithSuccessors for ReadOnlyBodyCache<'a, 'tcx> {
-impl Deref for ReadOnlyBodyCache<'a, 'tcx> {
-    type Target = Body<'tcx>;
+impl Deref for ReadOnlyBodyAndCache<'a, 'tcx> {
+    type Target = &'a Body<'tcx>;
```
Btw as I mentioned on the original PR, I've made this change in #65947, so I'd prefer landing that first.

Nashenas88 replied:
I saw that. I didn't plan to merge this until that one was merged first. I just wanted to make that change in this branch too so I could verify that the other changes wouldn't cause any issues. I'll update the main detail of the PR to make note of that.

bors commented:
:umbrella: The latest upstream changes (presumably #65947) made this pull request unmergeable. Please resolve the merge conflicts.

Nashenas88 commented:
@eddyb I rebased the PR and updated the main comment to reflect the minimized changes. Let me know if this still makes sense to move forward, otherwise I can just close this PR.

bors commented:
:umbrella: The latest upstream changes (presumably #66815) made this pull request unmergeable. Please resolve the merge conflicts.

oli-obk commented:
@bors r=eddyb

bors commented:
:pushpin: Commit a5e144b954df6a4e2db974a292a3328b100fd41c has been approved by `eddyb`

<!-- @bors r=eddyb a5e144b954df6a4e2db974a292a3328b100fd41c -->
<!-- homu: {"type":"Approved","sha":"a5e144b954df6a4e2db974a292a3328b100fd41c","approver":"eddyb"} -->
